lib/src/models/binned.dart

```diff
@@ -639,41 +686,127 @@ abstract class BinnedChartState<T extends BinnedChart> extends State<T>
-          SelectedBin selectedBin = SelectedBin(selectedBins!.seriesIndex, nextBin);
-          _updatedBinSelection(selectedBin);
+
+        // Detect Shift key release
+        if (isShiftKey(event.logicalKey)) {
+          isShiftKeyPressed = false;
         }
-      } else if (event.logicalKey == LogicalKeyboardKey.arrowUp) {
-      } else if (event.logicalKey == LogicalKeyboardKey.arrowDown) {}
-      setState(() {
-        scaleShiftKey = null;
       });
     }
-    return KeyEventResult.ignored;
+    return KeyEventResult.handled; // Indicate the event was handled
+  }
+
+  void _navigateBins(LogicalKeyboardKey key) {
+    if (selectedBins == null || selectedBins!.selectedBins.isEmpty) {
+      if (binContainers.isNotEmpty) {
+        final firstSeriesIndex = binContainers.keys.first;
+        if (binContainers[firstSeriesIndex]!.bins.isNotEmpty) {
+          final firstBin = SelectedBin(firstSeriesIndex, 0);
+          _updatedBinSelection(firstBin);
+        }
+      }
+      return;
+    }
+
+    final lastSelectedBinSeries = selectedBins!.selectedBins.keys.last;
+    final lastSelectedBinIndex = selectedBins!.selectedBins[lastSelectedBinSeries]!.last;
+
+    final currentSeries = binContainers[lastSelectedBinSeries]!;
+    int newBinIndex = lastSelectedBinIndex;
+
+    if (key == LogicalKeyboardKey.arrowLeft) {
+      if (newBinIndex == 0) {
+        return;
+      }
+      newBinIndex--;
+    } else if (key == LogicalKeyboardKey.arrowRight) {
+      if (newBinIndex == currentSeries.bins.length - 1) {
+        return;
+      }
+      newBinIndex++;
+    }
+
+    final newSelectedBin = SelectedBin(lastSelectedBinSeries, newBinIndex);
+    final Object thisSeriesIndex = newSelectedBin.seriesIndex;
+    final int thisBinIndex = newSelectedBin.binIndex;
+
+    if (isShiftKeyPressed) {
+      if (!selectedBins!.selectedBins.containsKey(thisSeriesIndex)) {
+        selectedBins!.addBin(thisSeriesIndex, thisBinIndex);
+      } else {
+        final existingSelection = selectedBins!.selectedBins[thisSeriesIndex]!;
+        int startIndex = existingSelection.reduce((a, b) => a < b ? a : b);
+        int endIndex = existingSelection.reduce((a, b) => a > b ? a : b);
+        int newIndex = thisBinIndex;
+        if (newIndex > endIndex) {
+          // Expanding right
+          selectedBins!.addBin(thisSeriesIndex, newIndex);
+        } else if (newIndex < startIndex) {
+          // Expanding left
+          selectedBins!.addBin(thisSeriesIndex, newIndex);
+        } else {
+          // **Reversing direction: Contract selection**
+          if (newIndex > startIndex) {
+            // Moving right after selecting left → Shrink left boundary
+            selectedBins!.removeBin(thisSeriesIndex, startIndex);
+          } else if (newIndex < endIndex) {
+            // Moving left after selecting right → Shrink right boundary
+            selectedBins!.removeBin(thisSeriesIndex, endIndex);
+          }
+        }
+      }
+    } else {
```

Comment:
This is the same behavior that should occur if the user shift clicks, so it could be its own function and both shift clicking and shift arrow navigation can call it.